### ISSUE TITLE
How to control the map scale (Generated Track Map)?

### ISSUE THREAD

**Issue by Shaftoe62** [SHWotever/SimHub#823]
Ever since I started with SimHub I'm having difficulties getting the scaling of the (dynamic) Track Maps right.

I want to have the scale of every map (Dirt Rally 1&2) the same. So 100 meters on one map is the same as on any another map.

But somehow the scaling is different for every map.

The scaling somehow seems to be dependent on the size of the map. 

I've been trying to compensate for that by deriving a scling factor from the map recordings but I can't figure out how to apply that correctly on the BaseMapScale because:

1. I don't know what the benchmark (default) map size/scale is from where to start.
2. I don't know what the BaseMapScale range is. It seems to go from 1 to 6 but I have no way of knowing that for sure.

**Comment by SHWotever** [comment]
Hi ! Base scale is automatically computed based on a couple of factors :

For the static map :
```
 g.Margin = (int)(Math.Max(OpponentStyle.DotRadius + OpponentStyle.DotBorderThickness, PlayerStyle.DotRadius + PlayerStyle.DotBorderThickness) + this.TrackBorderWidth + this.TrackWidth);

            var scalex = (this.Width - 2 * g.Margin) / (double)g.CoordinatesRangeX;
            var scaley = (this.Height - 2 * g.Margin) / (double)g.CoordinatesRangeY;

           effectiveScale = Math.Min(scaley, scalex);
            
```
For the dynamic map (rotating) :
``` 
effectiveScale = Math.Min(BaseMapScale, Math.Max(3000 / (double)g.CoordinatesRangeX, 3000 / (double)g.CoordinatesRangeX));
```
Margin is hard coded (50px)
CoordinatesRangeX = max X of the record - min X of the record;
CoordinatesRangeY = max Y of the record - min Y of the record;

Dynamic map also includes a dynamic scaling based on speed I won't detail here
Overall there are probably other subilities and computations I can't grab them all as it's spread inside the rendering engine :(

**Comment by Shaftoe62** [comment]
For example:

These two maps, both long stages in New Zealand have a completely different scaling. 

The shown widht and height are of the red bounding boxes. So without any margin. 



![image](https://user-images.githubusercontent.com/53898212/118602040-6dda0b00-b7b2-11eb-8571-e816681fc7ba.png)
![image](https://user-images.githubusercontent.com/53898212/118602180-a548b780-b7b2-11eb-8cfb-d215df0f62ad.png)

**Comment by Shaftoe62** [comment]
Ok, thanks. My issue is with the dynamic/rotating map. I will study the formula(s) and see if I can figure out a way to compensate for the dynamic scaling by calculating the appropriate BaseMapScale. That must be possible.

But for that I need to know what the allowed input range for BaseMapScale is, if any...

**Comment by Shaftoe62** [comment]
Same for effectiveScale. Are there any restriction (max/min) to these values:

BaseMapScale
effectiveScale

**Comment by SHWotever** [comment]
No particular limitations.

The difficulty is that the code works in reverse from what you want. In simhub I precompute an expected scale as an input, add some fixed margins, add some margin to avoid the player dots to go out of the nap , make sure it stays within a reasonable size, then finally render the image, the result does not not have a real precise expected scale (that wasn't the goal at all), the output of the renderer only gives information about how to plot player/opponents on it.
After this on the dynamic map there is a dynamic unzoom feature entering the equation. Overall that makes way too many variables to reproduce it accurately.

**Comment by Shaftoe62** [comment]
Well. I 'll give it a try anyway :-)

If I read the formula correctly,

this part:

Math.Max(3000 / (double)g.CoordinatesRangeX, 3000 / (double)g.CoordinatesRangeX)

compares two equal numbers? Or should one of them be CoordinatesRangeY?

As it is now it looks redundant to me?

Is the same as:

effectiveScale = Math.Min(BaseMapScale, 3000 / (double)g.CoordinatesRangeX);

**Comment by Shaftoe62** [comment]
And, can I conclude that I can achieve what I want only if I take a BaseMapScale value that is equal or lower than the result of 3000 divided by the largest width of all maps ?

**Comment by Shaftoe62** [comment]
Ok, I've taken the largest dimensions of all the maps (including the Pikes Peak stages of DR1):

Largest Map Width: 9331.15
Largest Map Height: 9313.42

That gives me a fixed BaseMapScale of 0.32. And that works! Hurrah!

I've tried a Full Course map en a Sprint stage (which gave the largest differences) and the show the same scale:

Full Course:
![image](https://user-images.githubusercontent.com/53898212/118613417-06768800-b7bf-11eb-8d47-7b34d9e252b8.png)


Sprint:
![image](https://user-images.githubusercontent.com/53898212/118613065-ad0e5900-b7be-11eb-8724-14fe9b709018.png)

**Comment by SHWotever** [comment]
That looks close to it ;) Indeed as the map is dynamically moved in the map rectangle there is no need of margins everywhere etc.
I've fixed the Math.Max(3000 / (double)g.CoordinatesRangeX, 3000 / (double)g.CoordinatesRange**X**) it was clearly a typo ;)

**Comment by Shaftoe62** [comment]
I took the margins into account and then floored the result just to make sure the BaseMapScale value will always be the lowest number. It works fine. Of course it's a very small scale but in a rally that has it's advantages.

Also, there is no need to zoom out with a scale like this.

So, thanks again. Now I understand how it works and I'm totally happy with the solution.

**Comment by Shaftoe62** [comment]
I don't know if you read comments in closed issues but I have more thing on this:

Now the scale is fixed I can use the player dot size to create a circle of 150 meters around the dot. Which is the magic number for me to prepare to brake/shift etc when nearing a curve.

![image](https://user-images.githubusercontent.com/53898212/118831292-f1cde900-b8bf-11eb-8899-d34d5f39c969.png)


The circle you see in the image has a DIAMETER of 93 (300 * 0.31)

But 93 is also the value that I had to enter in the field PlayerStyle -> Dot RADIUS

Just that you know, the value denotes the diameter, not the radius.

Not important at all but I noticed it now I needed a calculated size.
